### ISSUE TITLE
no-unused-vars argsIgnorePattern affecting "args": "after-used" since 3.6.0

### ISSUE THREAD

**Issue by danpoltawski** [eslint/eslint#7250]
**Tell us about your environment**

* **ESLint Version:** v3.6.0
* **Node Version:** v6.6.0
* **npm Version:** 3.10.7

**What parser (default, Babel-ESLint, etc.) are you using?** deafult

**Please show your full configuration:** https://github.com/moodle/moodle/blob/master/.eslintrc

**What did you do? Please include the actual source code causing the issue.**

When attempting to upgrade to eslint 3.6.0, we got some new errors reported by no-unused-vars which should have been ignored by `args: 'after-used'`. It turned out that argsIgnorePattern is now affecting the result of things which should be ignored by `"args": "after-used"`

I attempted to isolate the code down to the smallest test case:

Problem file:
```js
/* eslint 'no-unused-vars': ['error', {"args": "after-used", "argsIgnorePattern": "^baz$"}] */

(function(foo, bar, baz) {
        return baz;
})();
```

Output:
```sh
$ eslint problem.js 

/Users/danp/www/eslintest/problem.js
  3:16  error  'bar' is defined but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

Ok file:
```js
/* eslint 'no-unused-vars': ['error', {"args": "after-used"}] */

(function(foo, bar, baz) {
        return baz;
})();
```

**Comment by danpoltawski** [comment]
(Note: this was detected because we have our argsIgnorePattern too liberally matching by mistake)

**Comment by not-an-aardvark** [comment]
This was probably caused by https://github.com/eslint/eslint/pull/7143.

I think the issue is that we now check for the last non-ignored parameter (in your case, `bar`) rather than the last parameter itself (`baz`). However, we don't account for cases where an ignored parameter is used.

**Comment by platinumazure** [comment]
Confirmed in the online demo. I think this is something we should consider for a patch release, if possible.

**Comment by mysticatea** [comment]
Thank you for this issue.

Though I'd like to say this is intentional since `baz` is ignored, we should fix this issue to prevent a breaking change without major version up.
I will work on this today (JST).

By the way, I'm interested in the option which warns such access to ignored variables. Maybe `disallowUseIgnored`. Or a separated rule such as [no-use-ignored-vars](https://github.com/mysticatea/eslint-plugin/blob/master/docs/rules/no-use-ignored-vars.md).

**Comment by danpoltawski** [comment]
Thanks! As mentioned above - in our case it was caused by a misconfiguration of argsIgnorePattern which matched a lot of args, so I am not sure how much you should be concerned about this being a problem in practice.
